webknossos-datastore/app/com/scalableminds/webknossos/datastore/explore/NgffV0_4Explorer.scala

```diff
@@ -0,0 +1,157 @@
+import com.scalableminds.webknossos.datastore.datareaders.zarr._
+import com.scalableminds.webknossos.datastore.datavault.VaultPath
+import com.scalableminds.webknossos.datastore.models.VoxelSize
+import com.scalableminds.webknossos.datastore.models.datasource.LayerViewConfiguration.LayerViewConfiguration
+import com.scalableminds.webknossos.datastore.models.datasource.{Category, DataFormat}
+
+import scala.concurrent.ExecutionContext
+
+class NgffV0_4Explorer(implicit val ec: ExecutionContext) extends RemoteLayerExplorer with NgffExplorationUtils {
+
+  override def name: String = "OME NGFF Zarr v0.4"
+
+  override def explore(remotePath: VaultPath, credentialId: Option[String]): Fox[List[(ZarrLayer, VoxelSize)]] =
+    for {
+      zattrsPath <- Fox.successful(remotePath / NgffMetadata.FILENAME_DOT_ZATTRS)
+      ngffHeader <- parseJsonFromPath[NgffMetadata](zattrsPath) ?~> s"Failed to read OME NGFF header at $zattrsPath"
+      labelLayers <- exploreLabelLayers(remotePath, credentialId).orElse(Fox.successful(List[(ZarrLayer, VoxelSize)]()))
+
+      layerLists: List[List[(ZarrLayer, VoxelSize)]] <- Fox.serialCombined(ngffHeader.multiscales)(multiscale => {
+        for {
+          channelCount <- getNgffMultiscaleChannelCount(multiscale, remotePath)
+          channelAttributes = getChannelAttributes(ngffHeader.omero)
+          layers <- layersFromNgffMultiscale(multiscale, remotePath, credentialId, channelCount, channelAttributes)
+        } yield layers
+      })
+      layers: List[(ZarrLayer, VoxelSize)] = layerLists.flatten
+    } yield layers ++ labelLayers
+
+  private def layersFromNgffMultiscale(multiscale: NgffMultiscalesItem,
+                                       remotePath: VaultPath,
+                                       credentialId: Option[String],
+                                       channelCount: Int,
+                                       channelAttributes: Option[Seq[ChannelAttributes]] = None,
+                                       isSegmentation: Boolean = false): Fox[List[(ZarrLayer, VoxelSize)]] =
+    for {
+      axisOrder <- extractAxisOrder(multiscale.axes) ?~> "Could not extract XYZ axis order mapping. Does the data have x, y and z axes, stated in multiscales metadata?"
+      unifiedAxisUnit <- selectAxisUnit(multiscale.axes, axisOrder)
+      axisUnitFactors <- extractAxisUnitFactors(unifiedAxisUnit, multiscale.axes, axisOrder) ?~> "Could not extract axis unit-to-nm factors"
+      voxelSizeInAxisUnits <- extractVoxelSizeInAxisUnits(
+        multiscale.datasets.map(_.coordinateTransformations),
+        axisOrder) ?~> "Could not extract voxel size from scale transforms"
+      voxelSizeFactor = voxelSizeInAxisUnits * axisUnitFactors
+      nameFromPath = guessNameFromPath(remotePath)
+      name = multiscale.name.getOrElse(nameFromPath)
+      layerTuples <- Fox.serialCombined((0 until channelCount).toList)({ channelIndex: Int =>
+        for {
```

Comment:
some of these functions have gotten pretty big. Maybe it would make sense to extract some parts into separate named functions? Maybe the nested for comprehensions are a good candidate.

Reply:
Create createLayers